ready for review

Michael-Ivlev opened the pull request.

Michael-Ivlev committed:
ready for review

Michael-Ivlev committed:
Update README.md

talantbekov123 commented:
# Summary:

You've done a really good job! There are a couple of things that need to be done before your project can be accepted. Please take a look, I attached screenshots below comments where applicable. Hope they will be easy to fix.
 
Note: I may attach only one screen, but you need to find all similar cases inside the project and fix them to receive points.


## Backend

### Performance Criteria
- [x] <!--3.64--> The repository contains all the necessary infrastructure files:
  - [x] A  package.json file.
  - [x] An .editorconfig file.
  - [x] An .eslintrc file, which extends the airbnb-base configuration, and devDependencies required for the linter.
  ###no eslint file provided. Please try to run `npx eslint ./` to launch eslint check. All eslint errors must be fixed.
  - [x] A .gitignore file.
  - [x] Additionally, an exception for `_id` is added in .eslintrc. The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
- [x] <!--3.64--> No linting errors.
###no eslint file provided. Please try to run `npx eslint ./` to launch eslint check. All eslint errors must be fixed.
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
  - [x] An `npm run start` command that starts the server on `localhost:3000`.
  - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [x] <!--12.57--> When all dependencies are installed, the application starts with npm run dev without errors.
- [ ] <!--25.45--> The following routes function as described:
	- [x] A request to `GET /users/me` returns information about the user (email and name).
	- [x] `GET /articles` returns all articles saved by the user.
	- [x] `POST /articles` creates an article with the data passed inside the request body.
	- [x] `DELETE /articles/articleId` deletes the saved article using `_id`.4
	###https://snipboard.io/EsZU5I.jpg Please check owner. When the user tries to delete someone else's card, an error with a 403 status should be returned. If you've forgotten something, or have some doubts, you can check this reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403
	- [x] `POST /signup` creates a user with the data passed inside the request body.
	- [x] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--3.64--> API errors are handled:
	- [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
	- [x] Asynchronous handlers end with a `catch()` block.
	- [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> In production mode, the database address is taken from `process.env`.
- [x] <!--3.64--> Safe password storage has been implemented:
	- [x] Passwords are stored in an encrypted format.
	- [x] The API does not return a password hash to the client.
- [x] <!--3.64--> Data is validated before being added to the database.
- [x] <!--3.64--> Users can't delete saved cards from other user profiles.
###https://snipboard.io/EsZU5I.jpg Please check owner. When the user tries to delete someone else’s card, an error with a 403 status should be returned. If you’ve forgotten something, or have some doubts, you can check this reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403
- [x] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md. 
- [x] <!--3.64--> Storing the private key for creating a JWT is implemented correctly:
	- [x] For the production build, it is stored in an `.env` file, and this file should not be added to Git.
	- [x] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine and an error won't occur if there is no `.env` file present.

### Best Practices
- [x] <!--1.5--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--1.5--> Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
- [x] <!--1.5--> Logging is set up:
    ###https://snipboard.io/vWyL4D.jpg limiter is connected before the request logger. Requests rejected by the limiter will not be added to the request.log. It is necessary to connect the logger first in the chain, and then the limiter
	- [x] All requests and responses are logged to the `request.log` file.
	- [x] All errors are logged to the `error.log` file.
	- [x] Log files aren't added to the git repository.
- [x] <!--1.5--> Errors are handled by a centralized handler.
###https://snipboard.io/xRKeHj.jpg and here https://snipboard.io/JgMPa6.jpg Middleware violates the principle of centralized error handling. Instead of returning a response directly, it should throw an appropriate exception.
- [ ] <!--1.5--> Centralized error handling is described inside a separate module.
###https://snipboard.io/9rzG5g.jpg
- [x] <!--1.5--> In case of an error, the API returns a response status that matches the error type.
- [x] <!--1.5--> The server can be accessed via HTTPS using the domain specified in `README.md`. 
- [x] <!--1.5--> The application API is located on a domain with a name of the following format: name.zone/api (not just name.zone).
	- [x] Correct: news-explorer.tk/api
	- [x] Incorrect: news-explorer.tk
- [x] <!--1.5--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.

### Recommendations
- [x] <!--0.83--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--0.83--> The Helmet module is used to set security-related headers.
- [x] <!--0.83--> Configuration and constants are stored in separate files:
	- [x] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
	- [x] Application constants (response and error messages) are stored inside a separate file with constants.
	###https://snipboard.io/bqTN3m.jpg
- [x] <!--0.83--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [x] <!--0.83--> The rate limiter is configured in a separate file and imported into app.js.
- [x] <!--0.83--> The API is hosted on a separate subdomain, e.g. api.news-explorer.tk.

## Number of points: 58

Michael-Ivlev commented:
@talantbekov123 Hey, Thank you for the Review.
- The owner checking when the user tries to delete someone else's card is implemented in middlewares/permissions.js, 
When i check it in postman i get the proper 403 error 
- I cant get it how to move centralized error handling in to separate module

Michael-Ivlev committed:
after review

fix eslint
add constants for error text
add user valid error message for no owner article delete

talantbekov123 commented:
# Summary:

You've done a really good job!  Glad to say that your project has been accepted.


## Backend

### Performance Criteria
- [x] <!--3.64--> The repository contains all the necessary infrastructure files:
  - [x] A  package.json file.
  - [x] An .editorconfig file.
  - [x] An .eslintrc file, which extends the airbnb-base configuration, and devDependencies required for the linter.
  - [x] A .gitignore file.
  - [x] Additionally, an exception for `_id` is added in .eslintrc. The following rules are forbidden: `eslint-disable`, `eslint-disable-line`, and `eslint-disable-next-line`
- [x] <!--3.64--> No linting errors.
- [x] <!--3.64--> The scripts section of the package.json file contains the following:
  - [x] An `npm run start` command that starts the server on `localhost:3000`.
  - [x] An `npm run dev` command that starts the server on `localhost:3000` with hot reloading.
- [x] <!--12.57--> When all dependencies are installed, the application starts with npm run dev without errors.
- [x] <!--25.45--> The following routes function as described:
	- [x] A request to `GET /users/me` returns information about the user (email and name).
	- [x] `GET /articles` returns all articles saved by the user.
	- [x] `POST /articles` creates an article with the data passed inside the request body.
	- [x] `DELETE /articles/articleId` deletes the saved article using `_id`.4
	- [x] `POST /signup` creates a user with the data passed inside the request body.
	- [x] `POST /signin` returns a JWT when the correct email and password are passed in the request body.
- [x] <!--3.64--> All routes are protected with authorization, except for `/signin` and `/signup`.
- [x] <!--3.64--> User routes and article routes are described in separate files.
- [x] <!--3.64--> API errors are handled:
	- [x] If something is wrong with the request, the server returns a response with an error message and a corresponding status.
	- [x] Asynchronous handlers end with a `catch()` block.
	- [x] The API does not return standard database or Node.js errors.
- [x] <!--3.64--> In production mode, the database address is taken from `process.env`.
- [x] <!--3.64--> Safe password storage has been implemented:
	- [x] Passwords are stored in an encrypted format.
	- [x] The API does not return a password hash to the client.
- [x] <!--3.64--> Data is validated before being added to the database.
- [x] <!--3.64--> Users can't delete saved cards from other user profiles.
- [x] <!--3.64--> The server can be accessed via HTTPS using the domain specified in README.md. 
- [x] <!--3.64--> Storing the private key for creating a JWT is implemented correctly:
	- [x] For the production build, it is stored in an `.env` file, and this file should not be added to Git.
	- [x] In development mode (`process.env.NODE_ENV !== 'production'`), the code runs and works fine and an error won't occur if there is no `.env` file present.

### Best Practices
- [x] <!--1.5--> Asynchronous operations are implemented using promises or async/await.
- [x] <!--1.5--> Requests are validated before being passed to the controller. The body and (where applicable) headers and parameters are checked against the corresponding schemas. If a request doesn't match the schema, the processing is not passed to the controller and the client receives a validation error.
- [x] <!--1.5--> Logging is set up:
	- [x] All requests and responses are logged to the `request.log` file.
	- [x] All errors are logged to the `error.log` file.
	- [x] Log files aren't added to the git repository.
- [x] <!--1.5--> Errors are handled by a centralized handler.
- [ ] <!--1.5--> Centralized error handling is described inside a separate module.
###https://snipboard.io/9rzG5g.jpg
- [x] <!--1.5--> In case of an error, the API returns a response status that matches the error type.
- [x] <!--1.5--> The server can be accessed via HTTPS using the domain specified in `README.md`. 
- [x] <!--1.5--> The application API is located on a domain with a name of the following format: name.zone/api (not just name.zone).
	- [x] Correct: news-explorer.tk/api
	- [x] Incorrect: news-explorer.tk
- [x] <!--1.5--> All routes are connected to the `index.js` file, which is located in the `routes` folder, and `app.js` contains one main route handled by routes.

### Recommendations
- [x] <!--0.83--> For API errors, classes have been created to extend the Error constructor.
- [x] <!--0.83--> The Helmet module is used to set security-related headers.
- [x] <!--0.83--> Configuration and constants are stored in separate files:
	- [x] The Mongo server address and the private key for the JWT in development mode are stored inside a separate configuration file.
	- [x] Application constants (response and error messages) are stored inside a separate file with constants.
- [x] <!--0.83--> A rate limiter is set up: the number of requests from a single IP address is limited to a particular value in a given amount of time.
- [x] <!--0.83--> The rate limiter is configured in a separate file and imported into app.js.
- [x] <!--0.83--> The API is hosted on a separate subdomain, e.g. api.news-explorer.tk.

## Number of points: 99
